### PR TITLE
Port aggregate counter sink

### LINKED ISSUE #102: Port aggregate-counter as s-c-s sink
As a developer, I'd like to port *aggregate-counter* module from XD to s-c-s repo, so I can use it as *sink* module to build streaming pipeline.

**Acceptance:**
* *aggregate-counter* as sink module is available in *spring-cloud-stream-modules* repo 
* maven repo includes *aggregate-counter* module with respective group/artifact/version
* Module docs are included in ref. guide

### PR DESCRIPTION
**Pull request by ilayaperumalg**
- Add aggregate counter sink configuration/properties
  - Add aggregate counter repository (redis/memory based)
  - Add tests

This resolves #102

### CODE REVIEW

File: aggregate-counter-sink/src/main/java/org/springframework/cloud/stream/module/metrics/AggregateCounterSink.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	private AggregateCounterSinkProperties config;
+
+	@Autowired
+	private AggregateCounterRepository aggregateCounterRepository;
+
+	private DateTimeFormatter dateTimeFormatter;
+
+	@ServiceActivator(inputChannel = Sink.INPUT)
+	public void process(Message<?> message) {
+		Double increment = this.config.getIncrementExpression().getValue(message, Double.class);
```
Comment (ericbottard) [inline aggregate-counter-sink/src/main/java/org/springframework/cloud/stream/module/metrics/AggregateCounterSink.java]: Why `Double` if we're gonna use `long` on line 57?

File: aggregate-counter-sink/src/main/java/org/springframework/cloud/stream/module/metrics/AggregateCounterSink.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+		String counterName = this.config.getComputedNameExpression().getValue(message, CharSequence.class).toString();
+
+		if (this.config.getTimeField() == null) {
+			this.aggregateCounterRepository.increment(counterName, increment.longValue());
```
Comment (ericbottard) [inline aggregate-counter-sink/src/main/java/org/springframework/cloud/stream/module/metrics/AggregateCounterSink.java]: I would reduce the repository API surface and pass `now()` as a third parameter here

File: aggregate-counter-sink/src/main/java/org/springframework/cloud/stream/module/metrics/AggregateCounterSink.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+		}
+		else {
+			String timeStampValue = this.config.getTimeField().getValue(message, String.class);
+			if (dateTimeFormatter == null) {
```
Comment (ericbottard) [inline aggregate-counter-sink/src/main/java/org/springframework/cloud/stream/module/metrics/AggregateCounterSink.java]: Why use a field here? Moreover, there could be a race condition here

File: aggregate-counter-sink/src/main/java/org/springframework/cloud/stream/module/metrics/AggregateCounterSinkProperties.java [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+	 */
+	private Expression nameExpression;
+
+	/**
+	 * A SpEL expression (against the incoming Message) to derive the timestamp value.
+	 */
+	private Expression timeField;
+
+	/**
+	 *  Increment value for each bucket as a SpEL against the message
```
Comment (ericbottard) [inline aggregate-counter-sink/src/main/java/org/springframework/cloud/stream/module/metrics/AggregateCounterSinkProperties.java]: Additional space at start of javadoc and missing end dot.

File: aggregate-counter-sink/src/main/java/org/springframework/cloud/stream/module/metrics/AggregateCounterSinkProperties.java [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+	public void setName(String name) {
+		this.name = name;
+	}
+
+	public Expression getNameExpression() {
+		return nameExpression;
+	}
+
+	public Expression getComputedNameExpression() {
+		return (nameExpression != null ? nameExpression : new LiteralExpression(getName()));
```
Comment (ericbottard) [inline aggregate-counter-sink/src/main/java/org/springframework/cloud/stream/module/metrics/AggregateCounterSinkProperties.java:83]: Seems weird that `nameExpression` is also tested in `getName()`

File: aggregate-counter-sink/src/test/java/org/springframework/cloud/stream/module/AggregateCounterTests.java [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+/**
+ * Tests the aggregate-counter module.
+ *
+ * @author Eric Bottard
+ * @author Ilayaperumal Gopinathan
+ */
+@RunWith(SpringJUnit4ClassRunner.class)
+@SpringApplicationConfiguration(classes = AggregateCounterSinkApplication.class)
+@DirtiesContext
```
Comment (ericbottard) [inline aggregate-counter-sink/src/test/java/org/springframework/cloud/stream/module/AggregateCounterTests.java]: Does it, really? Given that we reset the repository, seems like we can re-user the AppContext

File: aggregate-counter-sink/src/test/java/org/springframework/cloud/stream/module/AggregateCounterTests.java [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+@IntegrationTest({"store=redis", "server.port=-1"})
+public abstract class AggregateCounterTests {
+
+	@Rule
+	public RedisTestSupport redisTestSupport = new RedisTestSupport();
+
+	private static final String AGGREGATE_COUNTER_NAME = "aggregate-counter-test.foo";
```
Comment (ericbottard) [inline aggregate-counter-sink/src/test/java/org/springframework/cloud/stream/module/AggregateCounterTests.java:61]: Not clear what the difference between these two names is.

File: common/spring-cloud-stream-modules-analytics/pom.xml [hunk 8]
```diff
@@ -32,5 +32,10 @@
 			<!-- projects that depend on scsm-analytics should choose explicitly which store(s) they want -->
 			<optional>true</optional>
 		</dependency>
+		<dependency>
+			<groupId>joda-time</groupId>
+			<artifactId>joda-time</artifactId>
+			<version>2.9.2</version>
```
Comment (ericbottard) [inline common/spring-cloud-stream-modules-analytics/pom.xml]: Doesn't boot provide a version somehow?

File: common/spring-cloud-stream-modules-analytics/src/main/java/org/springframework/cloud/stream/module/metrics/AggregateCounter.java [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+	}
+
+	public AggregateCounterResolution getResolution() {
+		return resolution;
+	}
+
+	@Override
+	public String toString() {
+		return "AggregateCount{" +
+				"name='" + name + '\'' +
```
Comment (ericbottard) [inline common/spring-cloud-stream-modules-analytics/src/main/java/org/springframework/cloud/stream/module/metrics/AggregateCounter.java]: Can put the closing single quote as part of the subsequent String

File: common/spring-cloud-stream-modules-analytics/src/main/java/org/springframework/cloud/stream/module/metrics/AggregateCounterReader.java [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+	 * @param name the counter to query
+	 * @param nCounts the number of data points to return
+	 * @param resolution the resolution at which the data should be returned (minute, hour, day, month)
+	 * @return an object containing an indexed array of the aggregate counts for the given query.
+	 */
+	AggregateCounter getCounts(String name, int nCounts, AggregateCounterResolution resolution);
+
+	/**
+	 * Query function to allow the counts for a specific interval to be retrieved for the given resolution.
+	 *
```
Comment (ericbottard) [inline common/spring-cloud-stream-modules-analytics/src/main/java/org/springframework/cloud/stream/module/metrics/AggregateCounterReader.java:41]: extra line

File: time-source/src/main/java/org/springframework/cloud/stream/module/time/TimeSourceProperties.java [hunk 11]
```diff
@@ -17,6 +17,7 @@
 package org.springframework.cloud.stream.module.time;
 
 import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.cloud.stream.module.DateFormat;
```
Comment (ericbottard) [inline time-source/src/main/java/org/springframework/cloud/stream/module/time/TimeSourceProperties.java]: import added but not used?

File: aggregate-counter-sink/src/main/java/org/springframework/cloud/stream/module/metrics/AggregateCounterSinkProperties.java [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+
+	/**
+	 * A SpEL expression (against the incoming Message) to derive the timestamp value.
+	 */
+	private Expression timeField;
+
+	/**
+	 *  Increment value for each bucket as a SpEL against the message
+	 */
+	private Expression incrementExpression = new SpelExpressionParser().parseExpression("1");
```
Comment (artembilan) [inline aggregate-counter-sink/src/main/java/org/springframework/cloud/stream/module/metrics/AggregateCounterSinkProperties.java]: A `new ValueExpression<>(1);` would be much better.
I guess in most cases this expression will be used by default.

File: aggregate-counter-sink/src/main/java/org/springframework/cloud/stream/module/metrics/AggregateCounterSink.java (hunk 3 above)
Comment (artembilan) [inline aggregate-counter-sink/src/main/java/org/springframework/cloud/stream/module/metrics/AggregateCounterSink.java]: :+1: 
if `DateTimeFormatter` isn't Thread Safe, should be better to use a new instance for each invocation, or `synchronized` around that `parseDateTime`.

From other side do we really need JodaTime for such a task ?

File: aggregate-counter-sink/src/main/java/org/springframework/cloud/stream/module/metrics/AggregateCounterSink.java [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+
+	@Autowired
+	private AggregateCounterRepository aggregateCounterRepository;
+
+	private DateTimeFormatter dateTimeFormatter;
+
+	@ServiceActivator(inputChannel = Sink.INPUT)
+	public void process(Message<?> message) {
+		Double increment = this.config.getIncrementExpression().getValue(message, Double.class);
+		String counterName = this.config.getComputedNameExpression().getValue(message, CharSequence.class).toString();
```
Comment (artembilan) [inline aggregate-counter-sink/src/main/java/org/springframework/cloud/stream/module/metrics/AggregateCounterSink.java]: What is the point to use `CharSequence` and then convert it into `String` anyway?

**Comment by ericbottard** [comment]
We need to review the while TimeZone handling of this (cf issues we encountered in the SmartGrid demo).

File: aggregate-counter-sink/src/main/java/org/springframework/cloud/stream/module/metrics/AggregateCounterSinkProperties.java (hunk 5 above)
Comment (ilayaperumalg) [inline aggregate-counter-sink/src/main/java/org/springframework/cloud/stream/module/metrics/AggregateCounterSinkProperties.java:83]: The `nameExpression` check in `getName()` is needed for the `isExclusiveOptions` assertion.

**Comment by ilayaperumalg** [comment]
Rebased and updated the PR after addressing the review comments.

**Comment by ilayaperumalg** [comment]
Created PR to update the `spring-cloud-dataflow` artifact registry populator: https://github.com/spring-cloud/spring-cloud-dataflow/pull/461

**Comment by ericbottard** [comment]
A couple of issues:

In effect, the default store is not redis. This may be an issue with other counters after all, but we must find a working solution for this. Moreover, using `@ConditionalOnProperty` with a hardcoded property is likely to cause problems when we introduce a prefix. It would be better to rely on the actual instance of `AggregateCounterSinkProperties` somehow.

When actually using redis (explicit `--store=redis`), these are the keys I get when testing. Note the extra dot in there:

```
127.0.0.1:6379> keys *
 1) "spring.cloud.artifact.task"
 2) "aggregate-counters..aggregate-counter.20160321"
 3) "spring.cloud.artifact.sink"
 4) "aggregate-counters.aggregate-counter"
 5) "aggregate-counters..aggregate-counter.2016"
 6) "aggregate-counters..aggregate-counter.2016032114"
 7) "metric_meta.aggregatecounters.aggregate-counter"
 8) "groups.foo.time"
 9) "spring.cloud.artifact.processor"
10) "aggregate-counters..aggregate-counter.years"
11) "groups.foo.http"
12) "aggregate-counters..aggregate-counter.201603"
```

File: aggregate-counter-sink/src/main/java/org/springframework/cloud/stream/module/metrics/AggregateCounterSinkProperties.java [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+	private Expression timeField;
+
+	/**
+	 * Increment value for each bucket as a SpEL against the message
+	 */
+	private Expression incrementExpression = new ValueExpression<>(1L);
+
+	private String dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'";
+
+	private DateTimeFormatter dateFormatter;
```
Comment (ericbottard) [inline aggregate-counter-sink/src/main/java/org/springframework/cloud/stream/module/metrics/AggregateCounterSinkProperties.java]: unused

**Comment by ilayaperumalg** [comment]
Pushed the changes that avoids using ConditionalOnProperty:
- Use configuration properties to instantiate the appropriate repositories
  - This assumes the `spring-data-redis` is on the classpath
- Apply the changes to counter, field-value-counter repository instantiations as well
- Fix the repo prefix for Aggregate counter

**Comment by ericbottard** [comment]
Rebased, squashed as b378da5

**Pull request closed without merging**